powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/builder/AbstractBatchMessageHandlerBuilder.java

```diff
@@ -0,0 +1,114 @@
+package software.amazon.lambda.powertools.batch.builder;
+
+import com.amazonaws.services.lambda.runtime.Context;
+import software.amazon.lambda.powertools.batch.handler.BatchMessageHandler;
+
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+
+/**
+ *
+ * An abstract class to capture common arguments used across all the message-binding-specific batch processing
+ * builders. The builders provide a fluent interface to configure the batch processors. Any arguments specific
+ * to a particular batch binding can be added to the child builder.
+ *
+ * We capture types for the various messages involved, so that we can provide an interface that makes
+ * sense for the concrete child.
+ *
+ * @param <T> The type of a single message in the batch
+ * @param <C> The type of the child builder. We need this to provide a fluent interface - see also getThis()
+ * @param <E> The type of the Lambda batch event
+ * @param <R> The type of the batch response we return to Lambda
+ */
+abstract class AbstractBatchMessageHandlerBuilder<T, C, E, R> {
+    protected BiConsumer<T, Throwable> failureHandler;
+    protected Consumer<T> successHandler;
+
+    /**
+     * Provides a success handler. A success handler is invoked once for
+     * each message after it has been processed by the user-provided
+     * handler.
+     * If the success handler throws, the item in the batch will be
+     * marked failed.
+     *
+     * @param handler The handler to invoke
+     */
+    public C withSuccessHandler(Consumer<T> handler) {
+        this.successHandler = handler;
+        return getThis();
+    }
+
+    /**
+     * Provides a failure handler. A failure handler is invoked once
+     * for each message after it has failed to be processed by the
+     * user-provided handler.
+     *
+     * @param handler The handler to invoke on failure
+     */
+    public C withFailureHandler(BiConsumer<T, Throwable> handler) {
+        this.failureHandler = handler;
+        return getThis();
+    }
+
+    /**
+     * Builds a BatchMessageHandler that can be used to process batches, given
+     * a user-defined handler to process each item in the batch. This variant
+     * takes a function that consumes a raw message and the Lambda context.
+     * Note:  If you don't need the Lambda context, use the variant of this function
+     * that does not require it.
+     *
+     * @param handler Takes a raw message - the AWS-SDK-shaped inner type of the batch - to process
```

Comment:
AWS-SDK-shaped ?? event library you mean. not SDK (maybe give an example `SQSMessage`)